File: src/2002_maximum_product_of_the_length_of_two_palindromic_subsequences.js
/**
 * @param {string} s
 * @return {number}
 * Time Complexity: O(2^N)
 * Space Complexity: O(2^N)
 */
var maxProduct = function (s) {
    const N = s.length;
    const first = new Array(1 << N).fill(0),
        last = new Array(1 << N).fill(0);
    for (let i = 0; i < N; i++) {
        for (let j = 1 << i; j < 1 << (i + 1); j++) {
            first[j] = i;
        }
    }
    for (let i = 0; i < N; i++) {
        for (let j = 1 << i; j < 1 << N; j += 1 << (i + 1)) {
            last[j] = i;
        }
    }
    const dp = Memo((m) => {
        if ((m & (m - 1)) === 0) {
            return m != 0;
        }
        const l = last[m],
            f = first[m];
        const lb = 1 << l,
            fb = 1 << f;
        return Math.max(
            dp(m - lb),
            dp(m - fb),
            dp(m - lb - fb) + (s[l] == s[f]) * 2,
        );
    });
    let ans = 0;
    for (let m = 1; m < 1 << N; m++) {
        ans = Math.max(ans, dp(m) * dp((1 << N) - 1 - m));
    }
    return ans;
};

var Memo = (func) => {
    const map = new Map();
    var wrapper = (m) => {
        if (!map.get(m)) {
            map.set(m, func(m));
        }
        return map.get(m);
    };
    return wrapper;
};

module.exports = { maxProduct };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(maxProduct("leetcodecom"), 9);
    assert.deepStrictEqual(maxProduct("bb"), 1);
    assert.deepStrictEqual(maxProduct("accbcaxxcxx"), 25);
}

File: src/2002_maximum_product_of_the_length_of_two_palindromic_subsequences.rs
struct Solution;

impl Solution {
    pub fn max_product(s: String) -> i32 {
        let s = s.into_bytes();
        let n = s.len();
        let limit = 1_usize << n;
        let mut first = vec![0_usize; limit];
        let mut last = vec![0_usize; limit];

        for i in 0..n {
            let start = 1_usize << i;
            let stop = 1_usize << (i + 1);
            for mask in start..stop {
                first[mask] = i;
            }
        }

        for i in 0..n {
            let step = 1_usize << (i + 1);
            let mut mask = 1_usize << i;
            while mask < limit {
                last[mask] = i;
                mask += step;
            }
        }

        let mut memo = vec![-1_i32; limit];
        let mut ans = 0;
        for mask in 1..limit {
            let product = Self::dp(mask, &s, &first, &last, &mut memo)
                * Self::dp(limit - 1 - mask, &s, &first, &last, &mut memo);
            ans = ans.max(product);
        }

        ans
    }

    fn dp(mask: usize, s: &[u8], first: &[usize], last: &[usize], memo: &mut [i32]) -> i32 {
        if mask == 0 {
            return 0;
        }
        if memo[mask] >= 0 {
            return memo[mask];
        }

        let value = if (mask & (mask - 1)) == 0 {
            1
        } else {
            let left = last[mask];
            let right = first[mask];
            let left_bit = 1_usize << left;
            let right_bit = 1_usize << right;
            let mut best = Self::dp(mask - left_bit, s, first, last, memo)
                .max(Self::dp(mask - right_bit, s, first, last, memo));
            let mut without_both = Self::dp(mask - left_bit - right_bit, s, first, last, memo);
            if s[left] == s[right] {
                without_both += 2;
            }
            best = best.max(without_both);
            best
        };

        memo[mask] = value;
        value
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mirrors_python_main_assertions() {
        assert_eq!(Solution::max_product(String::from("leetcodecom")), 9);
        assert_eq!(Solution::max_product(String::from("bb")), 1);
        assert_eq!(Solution::max_product(String::from("accbcaxxcxx")), 25);
    }
}
